pep-0xxx.rst

```diff
@@ -0,0 +1,1149 @@
+TLS implementations vary wildly in how they allow users to select trust stores.
+For this reason, we need to provide a model that assumes very little about the
+form that trust stores take.
+
+This problem is the same as the one that the Certificate and PrivateKey types
+need to solve. For this reason, we use the exact same model, by creating an
+opaque type that can encapsulate the various means that TLS backends may open
+a trust store.
+
+A given TLS implementation is not required to implement all of the
+constructors. However, it is strongly recommended that a given TLS
+implementation provide the ``system`` constructor if at all possible, as this
+is the most common validation trust store that is used. Concrete
+implementations may also add their own constructors.
+
+Concrete implementations should aim to have TrustStore objects be hashable if
+at all possible. This will help ensure that TLSConfiguration objects used with
+an individual concrete implementation are also hashable.
+
+    class TrustStore(metaclass=ABCMeta):
+        @abstractclassmethod
+        def system(cls) -> TrustStore:
+            """
+            Returns a TrustStore object that represents the system trust
+            database.
+            """
+
+        @abstractclassmethod
+        def from_pem_file(cls, path: Union[pathlib.Path, bytes, str]) -> TrustStore:
+            """
+            Initializes a trust store from a single file full of PEMs.
+            """
+
+
+Changes to the Standard Library
+===============================
+
+The portions of the standard library that interact with TLS should be revised
+to use these ABCs. This will allow them to function with other TLS backends.
+This includes the following modules:
+
+- asyncio
+- ftplib
+- http.client
+- imaplib
+- nntplib
+- poplib
+- smtplib
+
+
+Future
+======
+
+Major future TLS features may require revisions of these ABCs. These revisions
+should be made cautiously: many backends may not be able to move forward
+swiftly, and will be invalidated by changes in these ABCs. This is acceptable,
+but wherever possible features that are specific to individual implementations
+should not be added to the ABCs. The ABCs should restrict themselves to
+high-level descriptions of IETF-specified features.
+
+
+ToDo
+====
+
+* Consider adding a new parameter (``valid_subjects``?) to ``wrap_socket`` and
+  ``wrap_buffers`` that specifies in a *typed* manner what kind of entries in
+  the SAN field are acceptable. This would break the union between SNI and
+  cert validation, which may be a good thing (you can't SNI an IP address, but
+  you can validate a cert with one if you want).
+* It's annoying that there's no type signature for fileobj. Do I really have to
+  define one as part of this PEP? Otherwise, how do I define the types of the
+  arguments to ``wrap_buffers``?
+* Do we need ways to control hostname validation?
+* Do we need to support getpeercert? Should we always return DER instead of the
```

Comment:
OpenSSL can only return the chain from the TLS handshake. The chain usually lacks the trust anchor. In case it contains the root CA, that certificate is not trustworthy. AFAIK ```SSL_get_peer_cert_chain``` is just raw values. We would need to implement a verify callback to get the true chain. In case of SSL session resumption, the chain information is missing completely.